gluefactory/models/matchers/lightgluestick.py

```diff
@@ -0,0 +1,910 @@
+import warnings
+from pathlib import Path
+from typing import Callable, List, Optional, Tuple
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from omegaconf import OmegaConf
+from torch import nn
+from torch.utils.checkpoint import checkpoint
+
+from ...settings import DATA_PATH
+from ..base_model import BaseModel
+from ..utils.losses import NLLLoss
+from ..utils.metrics import matcher_metrics
+
+FLASH_AVAILABLE = hasattr(F, "scaled_dot_product_attention")
+
+torch.backends.cudnn.deterministic = True
+ETH_EPS = 1e-8
+
+
+@torch.cuda.amp.custom_fwd(cast_inputs=torch.float32)
+def normalize_keypoints(
+    kpts: torch.Tensor, size: Optional[torch.Tensor] = None
+) -> torch.Tensor:
+    if size is None:
+        size = 1 + kpts.max(-2).values - kpts.min(-2).values
+    elif not isinstance(size, torch.Tensor):
+        size = torch.tensor(size, device=kpts.device, dtype=kpts.dtype)
+    size = size.to(kpts)
+    shift = size / 2
+    scale = size.max(-1).values / 2
+    kpts = (kpts - shift[..., None, :]) / scale[..., None, None]
+    return kpts
+
+
+def rotate_half(x: torch.Tensor) -> torch.Tensor:
+    x = x.unflatten(-1, (-1, 2))
+    x1, x2 = x.unbind(dim=-1)
+    return torch.stack((-x2, x1), dim=-1).flatten(start_dim=-2)
+
+
+def apply_cached_rotary_emb(freqs: torch.Tensor, t: torch.Tensor) -> torch.Tensor:
+    return (t * freqs[0]) + (rotate_half(t) * freqs[1])
```

Comment:
If this is the exact same as in LightGlue, let's directly import from there maybe? The same applies for LearnableFourierPositionalEncoding, TokenConfidence, Attention, SelfBlock, CrossBlock, etc.